### PR TITLE
[refactor] Refactored and unified CC backend, removed CCProgram and use CCProgramImpl instead.

### PR DESCRIPTION
**Pull request by Leonz5288**
Sorry for the large PR, primary changes reside in `program.cpp`, `cc_program.cpp` and `cc_program.h`.

### CODE REVIEW

File: taichi/backends/cc/cc_kernel.h [hunk 1]
```diff
@@ -27,7 +30,7 @@ class CCKernel {
   }
 
  private:
-  CCProgram *program;
+  CCProgramImpl *cc_program_impl;
```
Comment (ljcc0930) [inline taichi/backends/cc/cc_kernel.h:33]:
```suggestion
  CCProgramImpl *cc_program_impl_;
```

File: taichi/backends/cc/cc_program.cpp [hunk 2]
```diff
@@ -1,18 +1,54 @@
...
-#include "struct_cc.h"
-#include "cc_program.h"
-#include "cc_runtime.h"
-#include "cc_kernel.h"
-#include "cc_layout.h"
-#include "cc_utils.h"
-#include "context.h"
+#include "taichi/backends/cc/cc_program.h"
+
+using namespace taichi::lang::cccp;
```
Comment (ljcc0930) [inline taichi/backends/cc/cc_program.cpp:3]: It's better to use `namespace cccp {`?

Comment (Leonz5288) [reply taichi/backends/cc/cc_program.cpp:3]: The APIs here are for ProgramImpl, which does not belongs to cccp scope. Moreover, this is consistent with other unified backends I believe.

File: taichi/program/program.h [hunk 3]
```diff
@@ -299,7 +299,7 @@ class Program {
  public:
 #ifdef TI_WITH_CC
   // C backend related data structures
-  std::unique_ptr<cccp::CCProgram> cc_program;
+  // std::unique_ptr<cccp::CCProgram> cc_program;
 #endif
```
Comment (ljcc0930) [inline taichi/program/program.h:303]: Just delete this?

Comment (Leonz5288) [reply taichi/program/program.h:303]: Sure, forgot this.

**Review by ljcc0930 — APPROVED** [review]
LGTM! It would be great if can add an underscore after the private variables.

File: taichi/backends/cc/cc_program.cpp [hunk 4]
```diff
@@ -1,18 +1,54 @@
...
+FunctionType CCProgramImpl::compile(Kernel *kernel, OffloadedStmt *offloaded) {
+  CCKernelGen codegen(kernel, this);
+  auto ker = codegen.compile();
+  auto ker_ptr = ker.get();
+  this->add_kernel(std::move(ker));
+  return [ker_ptr](Context &ctx) { return ker_ptr->launch(&ctx); };
+}
+
+void CCProgramImpl::materialize_runtime(MemoryPool *memory_pool,
+                                        KernelProfilerBase *profiler,
```
Comment (ailzhang) [inline taichi/backends/cc/cc_program.cpp:22]: ditto for profiler.

File: taichi/backends/cc/cc_program.h [hunk 5]
```diff
@@ -42,23 +68,30 @@ class CCProgram {
...
+  void add_kernel(std::unique_ptr<CCKernel> kernel);
+  void init_runtime();
+
   std::vector<std::unique_ptr<CCKernel>> kernels;
   std::unique_ptr<CCContext> context;
   std::unique_ptr<CCRuntime> runtime;
   std::unique_ptr<CCLayout> layout;
   std::unique_ptr<DynamicLoader> dll;
   std::string dll_path;
+  std::vector<char> args_buf;
```
Comment (ailzhang) [inline taichi/backends/cc/cc_program.h:91]: underscore suffix?

Comment (Leonz5288) [reply taichi/backends/cc/cc_program.h:91]: I'll modify these in another PR, otherwise, it would be a LOT of files change.

File: taichi/backends/cc/struct_cc.h [hunk 6]
```diff
@@ -27,7 +28,7 @@ class CCLayoutGen {
     line_appender.append(std::move(f), std::move(args)...);
   }
 
-  CCProgram *program;
+  CCProgramImpl *cc_program_impl_;
```
Comment (k-ye) [inline taichi/backends/cc/struct_cc.h:31]:
```suggestion
  CCProgramImpl *cc_program_impl_{nullptr};
```

File: taichi/program/program.cpp [hunk 7]
```diff
@@ -177,12 +175,9 @@ FunctionType Program::compile(Kernel &kernel, OffloadedStmt *offloaded) {
   TI_AUTO_PROF;
   FunctionType ret = nullptr;
   if (arch_uses_llvm(config.arch) || kernel.arch == Arch::metal ||
```
Comment (k-ye) [inline taichi/program/program.cpp:177]: TODO: can we create a function to do this check in another PR? Right now it's repeated 3 times

File: taichi/backends/cc/cc_runtime.h [hunk 8]
```diff
@@ -25,8 +25,7 @@ class CCRuntime {
   std::string source;
 
  private:
-  [[maybe_unused]] CCProgram *program;
-
+  CCProgramImpl *cc_program_impl_;
```
Comment (k-ye) [inline taichi/backends/cc/cc_runtime.h:28]:
```suggestion
  CCProgramImpl *cc_program_impl_{nullptr};
```

File: taichi/backends/cc/cc_program.h [hunk 9]
```diff
@@ -42,23 +68,30 @@ class CCProgram {
...
+  std::vector<char> root_buf;
+  std::vector<char> gtmp_buf;
+  uint64 *result_buffer;
```
Comment (k-ye) [inline taichi/backends/cc/cc_program.h:94]:
```suggestion
  uint64 *result_buffer{nullptr};
```

**Review by k-ye — APPROVED** [review]
LGTM!

File: taichi/program/program.cpp (hunk 7 above)
Comment (Leonz5288) [reply taichi/program/program.cpp:177]: True, also will change some other nits.

**Comment by ailzhang** [comment]
Merging to unblock followup improvements, macos build took 40mins on this PR (mainly building cuda_device class etc) which is a bit abnormal.

File: taichi/backends/cc/cc_program.cpp [hunk 10]
```diff
@@ -1,18 +1,54 @@
...
+  result_buffer = *result_buffer_ptr;
+}
+
+void CCProgramImpl::materialize_snode_tree(
+    SNodeTree *tree,
+    std::vector<std::unique_ptr<SNodeTree>> &,
+    std::unordered_map<int, SNode *> &,
+    uint64 *result_buffer) {
+  auto *const root = tree->root();
+  this->compile_layout(root);
```
[the remark this answers is not in the document]
Comment (Leonz5288) [reply taichi/backends/cc/cc_program.cpp:36]: Resolve in another PR

**Comment by k-ye** [comment]
Let's associate these PRs with #2736 :-)
